packages/client/test/rpc/eth/getBalance.spec.ts

```diff
@@ -54,27 +27,54 @@ tape(`${method}: ensure balance deducts after a tx`, async (t) => {
   // genesis address with balance
   const address = Address.fromString('0xccfd725760a68823ff1e062f4cc97e1360e8d997')
 
-  // construct tx
+  // verify balance is genesis amount
+  const genesisBalance = new BN(toBuffer('0x15ac56edc4d12c0000'))
+  let req = params(method, [address.toString(), 'latest'])
+  let expectRes = (res: any) => {
+    const msg = 'should return the correct genesis balance'
+    t.equal(res.body.result, bnToHex(genesisBalance), msg)
+  }
+  await baseRequest(t, server, req, 200, expectRes, false)
+
+  // construct block with tx
   const tx = Transaction.fromTxData({ gasLimit: 53000 }, { freeze: false })
   tx.getSenderAddress = () => {
     return address
   }
+  const block = Block.fromBlockData()
+  block.transactions[0] = tx
 
-  const { amountSpent } = await vm.runTx({ tx })
+  const result = await vm.runBlock({ block, generate: true, skipBlockValidation: true })
+  const { amountSpent } = result.results[0]
 
   // verify balance is genesis amount minus amountSpent
-  const genesisBalance = new BN(toBuffer('0x15ac56edc4d12c0000'))
   const expectedNewBalance = genesisBalance.sub(amountSpent)
-  const req = params(method, [address.toString(), 'latest'])
-  const expectRes = (res: any) => {
-    const msg = 'should return the correct balance'
-    const resultBN = new BN(toBuffer(res.body.result))
-    if (resultBN.eq(expectedNewBalance)) {
-      t.pass(msg)
-    } else {
-      throw new Error(msg)
-    }
+  req = params(method, [address.toString(), 'latest'])
+  expectRes = (res: any) => {
+    const msg = 'should return the correct balance after a tx'
+    t.equal(res.body.result, bnToHex(expectedNewBalance), msg)
   }
+  await baseRequest(t, server, req, 200, expectRes, false)
+
+  // verify we can query with "earliest"
+  req = params(method, [address.toString(), 'earliest'])
+  expectRes = (res: any) => {
+    const msg = "should return the correct balance with 'earliest'"
+    t.equal(res.body.result, bnToHex(genesisBalance), msg)
+  }
+  await baseRequest(t, server, req, 200, expectRes, false)
+
+  // verify we can query with a past block number
+  req = params(method, [address.toString(), '0x0'])
+  expectRes = (res: any) => {
+    const msg = 'should return the correct balance with a past block number'
+    t.equal(res.body.result, bnToHex(genesisBalance), msg)
+  }
+  await baseRequest(t, server, req, 200, expectRes, false)
+
+  // call with height that exceeds chain height
+  req = params(method, [address.toString(), '0x1'])
+  expectRes = checkError(t, INVALID_PARAMS, 'specified block greater than current height')
```

Comment:
This test in general looks great, but it does not explicitly check that when we pass a specific block, it actually uses that block. Probably just create 2 blocks (dont have to be valid) so we have blocks 0 (earliest), 1 and 2 (latest). Now you can check explicitly for earliest/latest (these are two different blocks), explicit 0x0, explicit 0x2, explicit 0x3 (throws) and explicit 0x1. Just check the balance of coinbase of all blocks should be fine.

Also 2 questions: (1) what happens if we pass a negative number over RPC? and (2) in general what happens if the RPC throws? Does it then still return a request with an error code, or does it fail to reply to the request?

Reply:
good idea, will add another block, thanks.

1. numbers are only passed as hex encoded over the RPC layer, so something like `-0x10`?

Trying with `toBuffer` which we mostly use internally to format, I get:
> Error: Cannot convert string to buffer. toBuffer only supports 0x-prefixed hex strings and this string was given: -0x10

2. the RPC needs to throw an error of shape `{ code, message }` to properly transmit as an JSON-RPC formatted error object. otherwise if something throws deeper, sometimes it may return as `Internal Error` with no error object/msg displayed, which is annoying and challenging to debug :) so you can see in some delicate places in the engine API I do:

```typescript
} catch (error: any) {
      throw {
        code: INTERNAL_ERROR,
        message: error.message.toString(),
      }
}
```

Reply:
Ah great thanks.

For (2) I think we should log the entire error message in the console? In Geth, for instance, if you do an RPC call and there is some internal error (e.g. trying to mine without etherbase/coinbase set) then it puts `[WARN] <error message>` in the console.

Reply:
(This is not in the scope of this PR though)

Reply:
yeah I agree with you, haven't found a straightforward way to forward all errors like this through `jayson` yet, definitely worth more investigation to see if it is possible (i assume it should be, we have to place a catch at a higher level and throw with the right obj shape)